File: leetcode_17_Letter Combinations of a Phone Number_recursive.cpp
class Solution {
public:
    vector<string> vec = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    vector<string> letterCombinations(string digits) {
        vector<string> ans;
        backtracking(ans, "", 0, digits);
        return ans;
    }
    void backtracking(vector<string>& ans, string str, int index, string& digits){
        if(str.size() == digits.size()){
            if(str.size() != 0) ans.push_back(str);
            return;
        }
        int id = digits[index] - '0';
        for(int i = 0; i < vec[id].size(); i++){
            backtracking(ans, str + vec[id][i], index + 1, digits);
        }
    }
};

File: leetcode_17_Letter Combinations of a Phone Number_recursive.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        m = {"2":"abc", "3":"def", "4":"ghi", "5":"jkl", "6":"mno", "7":"pqrs", "8":"tuv", "9":"wxyz"}
        ans = []
        if len(digits) == 0: return []
        if len(digits) == 1:
            return list(m[digits[0]])
        else:
            result = self.letterCombinations(digits[1:])
            for r in result:
                for j in m[digits[0]]:
                    ans.append(j + r)
        return ans